nimbleclient/v1/restclient.py

```diff
@@ -164,81 +168,93 @@ def close_connection(self):
-        """Wrapper for GET requests"""
+    def build_advanced_criteria(self, operator, criteria):
+        """Constructs advanced criteria and returns JSON object."""
+
+        return {
+            'data': {
+                '_constructor': 'AdvancedCriteria',
+                'operator': operator,
+                'criteria': criteria,
+            },
+            'operationType': 'fetch'
+        }
+
+    def get(self, endpoint, filter=None, **params):
+        """Wrapper for GET requests with filters and advanced criteria"""
+
+        url = f'https://{self.hostname}:{self.port}/{endpoint}'
+
+        logging.debug("Query Params: %s", json.dumps(params, indent=4))
 
-        url=f'https://{self.hostname}:{self.port}/{endpoint}'
         try:
+            # Init response data and query for more paginated data records if available
+            response_data = []
             while 1:
-                response = requests.get(
-                    url,
-                    params=params,
-                    headers=self._headers,
-                    verify=False
-                )
+                if filter is not None:
+                    response = requests.post(
+                        url,
+                        params=params,
+                        json=filter,
+                        headers=self._headers,
+                        verify=False
+                    )
+                else:
+                    response = requests.get(
+                        url,
+                        params=params,
+                        headers=self._headers,
+                        verify=False
+                    )
 
                 if response.status_code >= 400:
                     if 'SM_http_unauthorized' in str(response.content):
                         self._refresh_connection()
                     else:
-                        print("URL: ", url)
                         raise NimOSAPIError(response.json())
-                else:
+
+                response = response.json()
+                logging.debug("API RESPONSE: %s", json.dumps(response, indent=4))
+
+                # Check for errors if any in the response (Treat partial response as an error)
+                if 'messages' in response:
+                    raise NimOSAPIError(response['messages'])
+
+                # Get response
+                if 'totalRows' not in response:
+                    logging.debug("Returning Data: %s", json.dumps(response, indent=4))
+                    return response
+
+                # Append to the resultant data
+                response_data.extend(response["data"])
+
+                # When pageSize is requested, then return only those no. of records
+                if 'pageSize' in params:
+                    break
+
+                # When end_row is requested, then return only upto 'end_row' records
+                if 'endRow' in params and params['endRow'] == response['endRow']:
+                    break
```

Comment:
You shouldn't need this logic. Follow what the response gives you. In other words, if the REST response says there's more data, go get it until there is no more data.

Reply:
This was added to be inline with REST API behavior. If the user specifies `endRow`, then we should **stop** retrieving further records even if there are more records to process. Thats the expectation of **endRow** filter.

**API REQUEST:**
```
https://suneeth-vm1.vlab.nimblestorage.com:5392/v1/volumes?endRow=3
```

**API RESPONSE:**
```
{
    "startRow": 0,
    "endRow": 3,
    "totalRows": 443,
    "data": [
        {
            "id": "0601a32bf8f45646a5000000000000000000000222",
            "name": "pvc-0c051a49-5af4-11ea-9000-005056966823"
        },
        {
            "id": "0601a32bf8f45646a5000000000000000000000530",
            "name": "pvc-325bdb4d-6f5a-11ea-9000-005056966823"
        },
        {
            "id": "0601a32bf8f45646a50000000000000000000004b5",
            "name": "pvc-dcb0468b-6b6c-11ea-9000-005056966823"
        }
    ]
}
```

Reply:
The endRow conditional in particular is not really the problem, it's the paging logic.

First thing to note is that the NimOS REST API will **not** do paging by default, and I'm not sure why we would even want to support that in this client. The whole point is to make it easier for someone to use the API.

Let's say someone passes a pageSize of 5. This code would sit here and issue 443 / 5 = 89 separate HTTP requests to retrieve what you could've pulled with just 1 request that contained all the responses. There's no way that 89 separate requests will be more efficient than the 1. In fact, it's **highly** likely that you'll error out before request number 20 or so because the array can't handle more than 8 requests at the same time.

If you still want to support that capability, then the way the logic is written right now, you'll only ever retrieve the one page, because you break out of the loop as soon as the pageSize parameter is present (regardless of its value).

Separately, once you specify pageSize, there's no point in doing math to figure out what the next page is. The response tells you the next URL you should call in the `paging` field (see below). 

```
{'data': [{'id': '0601a32bf8f45646a5000000000000000000000008',
           'name': 'pvc-92c6a7c5-4c2f-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a5000000000000000000000009',
           'name': 'pvc-d8199f97-4c2f-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000000a',
           'name': 'pvc-1b6846d3-4c30-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000000b',
           'name': 'pvc-61f352ea-4c30-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000000d',
           'name': 'pvc-f27aaba5-4c30-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000000e',
           'name': 'pvc-fba983ce-4c30-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000000f',
           'name': 'pvc-37eb8495-4c31-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000001d',
           'name': 'pvc-88695760-4c4f-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000001e',
           'name': 'pvc-cb520459-4c4f-11ea-9000-005056966823'},
          {'id': '0601a32bf8f45646a500000000000000000000001f',
           'name': 'pvc-11994d55-4c50-11ea-9000-005056966823'}],
 'endRow': 10,
 'paging': {'current_cursor_value': '0601a32bf8f45646a500000000000000000000001f',
            'cursor': 'id',
            'next_url': '/v1/volumes?pageSize=10&id%3E0601a32bf8f45646a500000000000000000000001f'},
 'startRow': 0,
 'totalRows': 444}
```

So back to the original point about `endRow`. The API responds with all the data you asked for and if you're using pageSize, it tells you that there's more data to get with the `paging` response attribute. That means that the logic is insanely simple:
* GET data
* Does data contain `paging`?
  * YES, then GET `response['paging']['next_url']`
  * NO, you're done

There's no internal state to track or math to do.

Reply:
We should not be making 89 requests in order to get all of the rows when a user specifies a pageSize of 5.  Rather, if the pageSize is specified we should only ever retrieve the number of rows specified in the pageSize.  If startRow is also populated, then we will retrieve pageSize rows starting at startRow.